xarray/backends/common.py

```diff
@@ -503,3 +503,43 @@ def assert_open(self):
         if not self._isopen:
             raise AssertionError('internal failure: file must be open '
                                  'if `autoclose=True` is used.')
+
+
+class PickleByReconstructionWrapper(object):
+
+    def __init__(self, opener, *args, **kwargs):
+
+        self.opener = opener
+        self.open_args = args
+        self.open_kwargs = kwargs
+
+        self._ds = None
+        self._isopen = False
+
+    @property
+    def value(self):
```

Comment:
I'm a little concerned about the complexity of making file opening lazy. State makes things harder to reason about. For example, what happens with the `mode` argument if never actually accessed the file?

For now can we stick with constructing `value` eagerly?

Reply:
I'm actually not a big fan of dropping the lazy opening. For one, there will be a non-negligible performance penalty. Perhaps a better argument is that we've been using this logic successfully in xarray for a while now. That said, its not clear to me what the potential problems with this approach so I'm open to convincing otherwise.

Reply:
Don't we always open files immediately to pull out metadata? (e.g., variables and dimensions)